lib/tom_queue/deferred_work_manager.rb

```diff
@@ -113,6 +121,24 @@ def stop
       deferred_set && deferred_set.interrupt
     end
 
-  end
+    # Register handlers to shut down the manager gracefully
+    def register_signal_handlers
+      Thread.main[:signals] = []
+
+      [:QUIT, :INT, :TERM].each do |sig|
+        trap(sig) do
+          Thread.main[:signals] << sig
+        end
+      end
```

Comment:
As mentioned, there is a much simpler way to do this by signalling the condition-variable to interrupt the blocking pop operation immediately, having set a `@shutdown` flag.